jc/parsers/ascii.py

```diff
@@ -0,0 +1,178 @@
+
+    [
+      {
+        "column1":     string,
+        "column2":     boolean,
+        "column3":     integer
+      }
+    ]
+    
+Examples:
+
+    $ virsh domifaddr foo | jc --ascii
+    [{"name":"vnet0","mac_address":"01:23:45:67:89:ab","protocol":"ipv4","address":"192.168.122.100/24"}]
+
+"""
+import jc.utils
+import jc.parsers.universal
+import re
+
+
+class info():
+    """Provides parser metadata (version, author, etc.)"""
+    version = '1.0'
+    description = 'ASCII table parser'
+    author = 'Alexandru Barbur'
+    author_email = 'alex@ctrlc.name'
+    # details = 'enter any other details here'
+
+    # compatible options: linux, darwin, cygwin, win32, aix, freebsd
+    compatible = ['linux', 'darwin', 'cygwin', 'win32', 'aix', 'freebsd']
+    magic_commands = ['ascii']
+
+
+__version__ = info.version
+
+
+def _process(proc_data):
+    """
+    Final processing to conform to the schema.
+
+    Parameters:
+
+        proc_data:   (List of Dictionaries) raw structured data to process
+
+    Returns:
+
+        List of Dictionaries. Structured to conform to the schema.
+    """
+
+    # rebuild output for added semantic information
+    # use helper functions in jc.utils for int, float, bool conversions and timestamps
+
+    return proc_data
+
+
+def _snake_case(value):
+    """
+    Convert a value to snake case
+    
+    Parameters:
+    
+        value:    (string) original value
+        
+    Returns:
+    
+        Snake cased value.
+    """
+
+    value = value.replace('-', '_')
+    value = re.sub(r'([A-Z]+)([A-Z][a-z])', r'\1_\2', value)
+    value = re.sub(r'([a-z\d])([A-Z])', r'\1_\2', value)
+    value = re.sub(r'\s+', r'_', value)
+    return value.lower()
+
+
+def _process_headers(line):
+    """
+    Detect and clean header values in the table header row
+    
+    Parameters:
+    
+        line:    (string) header row line data
+        
+    Returns:
+    
+        Line with cleaned header values.
+    """
+
+    # XXX: what about tables that only use a single space between columns?
+    matches = list(re.finditer(r"\w+( \w+)*", line))
+    for match in matches:
+        original_header = match.group()
+
+        # snake case the header and pad or trim it to match original length
+        clean_header = _snake_case(original_header)
+        clean_header.ljust(len(original_header))
+        clean_header = clean_header[:len(original_header)]
+
+        # replace the matched header with the cleaned version in the line
+        line = line[:match.start()] + clean_header + line[match.end():]
+
+    return line
+    
+
+def _header_or_data_row(line):
+    """
+    Filter lines of text for header or data rows
+    
+    Parameters:
+    
+        line:    (string) line of text
+        
+    Returns:
+    
+        True if the line contains headers or data
+    """
+
+    # ignore empty or blank lines
+    if not line:
+        return False
+
+    # ignore separator lines (i.e. "----")
+    if len(set(line)) <= 1:
```

Comment:
Too clever or too simple?

Reply:
Ha - I really like that. Very clever and elegant!

Reply:
Well good :)

Reply:
Actually, maybe just one little tweak:
```
if len(set(line.replace(' ', ''))) <= 1:
```
This way it doesn't count spaces as part of the separator line. This is for cases like the output of `jtbl`:
```
$ jc df | jtbl -n
filesystem       512_blocks       used    available    iused                ifree  mounted_on           capacity_percent    iused_percent
-------------  ------------  ---------  -----------  -------  -------------------  -----------------  ------------------  ---------------
/dev/disk1s1      976490576  664388168    300216728  2679309  9223372036852096498  /                                  69                0
devfs                   660        660            0     1142                    0  /dev                              100              100
/dev/disk1s4      976490576   10485984    300216728        5  9223372036854775802  /private/var/vm                     4                0
```

Reply:
What if there's a table like this:

```
Column  Column
--------  ---------
-                -
X               X
```

Although I suppose you could argue that it's ok to exclude those rows.

It feels like there's no way to solve this problem for the general case when you're looking at one row at a time. To handle that we probably need to:

1. Looks at all the columns and picks the separators (based on matching values in each row) then determine the valid text range for each.
2. Looks at all the rows and picks out the separators (based on matching values in each column possibly ignoring the ranges used to separate columns as detected above) and keep the header and data rows.

The result of that could probably be fed into `sparse_table_parse` actually although we'd be reading the data at least 3 times then.

Reply:
Yeah, to do it right is going to be a bit of analysis. Part of me is cool with an MVP and let's see if any issues are reported. But it would be slick to do it right with a non-lossy parser. A lot more code, though.

Reply:
Another thought to keep things simple - only check for the separator line right after the header if one is detected. Still wouldn't be perfect, but might cover 80%+ of the use cases.